internal/dcache/rpc/client/client_pool.go

```diff
@@ -259,6 +340,136 @@ func (cp *clientPool) closeRPCClient(client *rpcClient) error {
 	return nil
 }
 
+// deleteRPCClient deletes an RPC client from the pool.
+// It first closes the client and then removes it from the pool.
+// This is used when the client is no longer needed, e.g. when the node is down and we want to
+// remove all clients to the node.
+//
+// NOTE: Caller MUST hold the clientPool lock.
+func (cp *clientPool) deleteRPCClient(client *rpcClient) error {
+	log.Debug("clientPool::deleteRPCClient: Deleting RPC client to %s node %s",
+		client.nodeAddress, client.nodeID)
+
+	common.Assert(cp.isLockHeld())
+
+	// Close the client first.
+	err := cp.closeRPCClient(client)
+	if err != nil {
+		return err
+	}
+
+	ncPool, exists := cp.clients[client.nodeID]
+	_ = exists
+	common.Assert(exists, client.nodeID)
+
+	//
+	// deleteRPCClient() MUST be called after removing client from the client pool, so
+	// the pool must have space for a new client.
+	//
+	common.Assert(len(ncPool.clientChan) < int(cp.maxPerNode),
+		len(ncPool.clientChan), cp.maxPerNode)
+
+	//
+	// Must only delete an active client.
+	// Also, clients which are deleted are not released, so we drop the numActive here, after
+	// closing the connection successfully above.
+	//
+	common.Assert(ncPool.numActive.Load() > 0)
+	ncPool.numActive.Add(-1)
+	return nil
+}
+
+// Delete all connections in the client pool corresponding to 'client'. This client would have been allocated
+// using a prior call to getRPCClient().
+// This is used for draining connections in the connection pool in case there is timeout error while
+// making an RPC call to the target node.
+// It closes the passed in connection and all existing connections in the pool, and if there are no
+// active connections and no connections in the channel, it deletes the node client pool.
+func (cp *clientPool) deleteAllRPCClients(client *rpcClient) error {
+	//
+	// deleteAllRPCClients() will be called only when we know for sure that an RPC request made using 'client'
+	// failed with a "timeout" error, we delete that client and all others in the pool.
+	//
+	cp.acquireLock()
+	defer cp.releaseLock()
+
+	numConnDeleted := 0
+	ncPool, exists := cp.clients[client.nodeID]
+	_ = exists
+
+	// client is allocated from the pool, so pool must exist.
+	common.Assert(exists)
+	common.Assert(ncPool.nodeID == client.nodeID, ncPool.nodeID, client.nodeID)
+
+	//
+	// Clients present in the pool. If it's less than cp.maxPerNode, rest are currently allocated to other
+	// callers. We cannot replenish those. Those will be deleted by their respective caller when their RPC
+	// requests fail with "timeout" error.
+	//
+	numClients := len(ncPool.clientChan)
+	common.Assert(numClients < int(cp.maxPerNode), numClients, cp.maxPerNode)
+
+	//
+	// Delete this client. This closes this client and removes it from the pool.
+	// It can only fail if Thrift fails to close the connection.
```

Comment:
Add
This should technically not happen, so we assert.

then add an assert in the below block

common.Assert(false, err)

Reply:
Done